### PR TITLE
Improve slow torchmetrics.Accuracy

### LINKED ISSUE #485: `torchmetrics.Accuracy` is too slow than normal pytorch code
### To Reproduce

```python
import torch
import torchmetrics
x = torch.rand(256,10)
y = torch.randint(10,(256,))

def MyAccuracy(output, target, topk=(1,)):
    """ Computes the precision@k for the specified values of k """
    maxk = max(topk)
    batch_size = target.size(0)

    _, pred = output.topk(maxk, 1, True, True)
    pred = pred.t()
    correct = pred.eq(target.reshape(1, -1).expand_as(pred))

    res = []
    for k in topk:
        correct_k = correct[:k].reshape(-1).float().sum(0, keepdim=True)
        res.append(torch.tensor(correct_k.mul_(100.0 / batch_size)))
    if len(res) == 1:
        return res[-1]
    return res
```

timing

```python
import timeit
def f1():
    MyAccuracy(x,y)
def f2():
    torchmetrics.Accuracy()(x,y)
def f3():
    torchmetrics.functional.accuracy(x,y)

timeit.timeit(f1, number=1000)
timeit.timeit(f2, number=1000)
timeit.timeit(f3, number=1000)
```

The result is on CPU

image

### Environment

- PyTorch Version (e.g., 1.0):
- OS (e.g., Linux):
- How you installed PyTorch (`conda`, `pip`, source):
- Build command you used (if compiling from source):
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by puhuk**
To resolve issue #485

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

Fixes #485.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
:]
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/functional/classification/accuracy.py [hunk 1]
```diff
@@ -388,31 +388,34 @@ def accuracy(
...
+            correct, total = _subset_accuracy_update(preds, target, threshold, top_k)
+            return _subset_accuracy_compute(correct, total)
+        tp, fp, tn, fn = _accuracy_update(
+            preds, target, reduce, mdmc_average, threshold, num_classes, top_k, multiclass, ignore_index, mode
+        )
+        return _accuracy_compute(tp, fp, tn, fn, average, mdmc_average, mode)
 
-    if num_classes and ignore_index is not None and (not 0 <= ignore_index < num_classes or num_classes == 1):
-        raise ValueError(f"The `ignore_index` {ignore_index} is not valid for inputs with {num_classes} classes")
+    except:
```
Comment (Borda) [inline torchmetrics/functional/classification/accuracy.py:405]: pls specify type of exception

**Review by Borda — CHANGES_REQUESTED** [review]

**Comment by justusschock** [comment]
@puhuk as you can see in https://dev.azure.com/PytorchLightning/Metrics/_build/results?buildId=41050&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=98354d77-e326-51ec-536f-1549451db1fa some of the tests failed with your adjustments

**Comment by puhuk** [comment]
@justusschock Let me check and resend

**Comment by Borda** [comment]
@puhuk how is it going here? :rabbit:

**Comment by puhuk** [comment]
@Borda Sorry for late submission of this PR. Because of some busy works in thesedays, I could not correct the commit. Could you give me some time by next weekend. If it is urgent, it is sure you can assign others to this :)

Sorry for lateness again.

**Comment by Borda** [comment]
> Could you give me some time by next weekend. If it is urgent, it is sure you can assign others to this :)

sure, no problem :rabbit: just let us know if need even more or other help...
cc: @SkafteNicki

**Review by Borda — APPROVED** [review]

**Review by justusschock — CHANGES_REQUESTED** [review]
Blocking to wait for @puhuk response and not accidentally merging.

**Comment by puhuk** [comment]
@Borda Sorry, but may I ask what is changed and why is it approved to merge ??

**Comment by Borda** [comment]
> Sorry, but may I ask what is changed and why is it approved to merge ??

apology, just overclocked instead of comment...

**Comment by Borda** [comment]
@puhuk how is it going here? :rabbit:

**Comment by puhuk** [comment]
@Borda,
Hi, sorry with my mistake. I thought it was approved and merged from PR and review history.
Sorry to bother you but may I ask what is remained with this PR?

**Comment by Borda** [comment]
> Hi, sorry with my mistake. I thought it was approved and merged from PR and review history. Sorry to bother you but may I ask what is remained with this PR?

check @justusschock comment https://github.com/PyTorchLightning/metrics/pull/609#pullrequestreview-815202418

**Comment by puhuk** [comment]
> > Hi, sorry with my mistake. I thought it was approved and merged from PR and review history. Sorry to bother you but may I ask what is remained with this PR?
> 
> check @justusschock comment [#609 (review)](https://github.com/PyTorchLightning/metrics/pull/609#pullrequestreview-815202418)

As I checked @justusschock 's comment, there is no review with code and before the comment, I reflected code reviews to the PR.
Could you kindly let me know what I'm missing now.

**Comment by Borda** [comment]
> Could you kindly let me know what I'm missing now.

- [ ] Do you have benchmarks on how much your implementation is faster?
- [ ] Fix the failing tests :rabbit:

**Comment by puhuk** [comment]
I leave benchmark as below.
It slightly improve current logic but hard to accomplish the time originally suggested by issue raiser (marsggbo from issue #485)

I tried to correct the failing test cases, but it seems hard for me to resolve and finish this PR.
May I ask to help with this.

```

# 0.161, Accuracy calculation logic of marsggbo (issue #485)
# 1.471  Current accuracy
# 0.297  torchmetrics.functional.accuracy
# 0.901  suggested PR logic

import torch
import torchmetrics
from metrics.torchmetrics.classification import Accuracy as AC

x = torch.rand(256,10)
y = torch.randint(10,(256,))
import timeit

def MyAccuracy(output, target, topk=(1,)):
    """ Computes the precision@k for the specified values of k """
    maxk = max(topk)
    batch_size = target.size(0)

    _, pred = output.topk(maxk, 1, True, True)
    pred = pred.t()
    correct = pred.eq(target.reshape(1, -1).expand_as(pred))

    res = []
    for k in topk:
        correct_k = correct[:k].reshape(-1).float().sum(0, keepdim=True)
        res.append(torch.tensor(correct_k.mul_(100.0 / batch_size)))
    if len(res) == 1:
        return res[-1]
    return res


def f1():
    MyAccuracy(x,y)
def f2():
    torchmetrics.Accuracy()(x,y)
def f3():
    torchmetrics.functional.accuracy(x,y)
def f4():
    # Test with changed accuracy function
    AC()(x,y)

print(timeit.timeit(f1, number=1000)) # 0.161, Accuracy calculation logic of marsggbo (issue #485)
print(timeit.timeit(f2, number=1000)) # 1.471  Current accuracy
print(timeit.timeit(f3, number=1000)) # 0.297  torchmetrics.functional.accuracy
print(timeit.timeit(f4, number=1000)) # 0.901  suggested PR logic

```

**Comment by SkafteNicki** [comment]
Closing PR as we in the near future is going to refactor the hole classification package (https://github.com/PyTorchLightning/metrics/projects/6) where we can better try to design all classification metrics (at least metrics that inherent from `StatScore`) to be faster.
Thanks for wanting to contribute @puhuk. Your code has made it clear that a large portion of the computational time is spent on input checking and that we should take this into account.
